scipy/linalg/basic.py

```diff
@@ -360,12 +360,38 @@ def solve_triangular(a, b, trans=0, lower=False, unit_diagonal=False,
                      (-info))
 
 
+def _diagonal_banded(l_and_u, a):
+    n = a.shape[1]
+    if a.shape != (n, n):
+        raise ValueError("Matrix must be square (has shape %s)" % (a.shape,))
+    (nlower, nupper) = l_and_u
+    if nlower > n or nupper > n:
+        raise ValueError("Number of nonzero diagonals must be less than square dimension")
+
+    upper = np.empty((nupper, n), dtype=a.dtype)
+    mid = np.empty((1, n), dtype=a.dtype)
+    lower = np.empty((nlower, n), dtype=a.dtype)
+
+    for i in range(1, nupper + 1):
+        for j in range(n - i):
+            upper[nupper - i, i + j] = a[j, i + j]
+
+    for i in range(n):
+        mid[0, i] = a[i, i]
+
+    for i in range(nlower):
+        for j in range(n - i - 1):
+            lower[i, j] = a[i + j + 1, j]
```

Comment:
Maybe just a longer-term thought, but if the linalg experts like this addition, I would suggest there's some opportunity for improved efficiency given three separate sets of Python loops, two of them nested. Maybe Cythonization at some point.